ops/testing.py

```diff
@@ -1544,6 +1583,93 @@ def evaluate_status(self) -> None:
+        the longest match takes precedence.
+        longest match takes precedence. The execution handler registration can be updated by
+        re-registering with the same command prefix.
+
+        The execution handler receives the timeout value in ExecArgs. If needed, it can raise a
+        TimeoutError to inform the harness of a timeout occurrence.
+
+        If :meth:`ops.Container.exec` is called with ``combine_stderr=True``, the execution
+        handler should, if required, weave the simulated standard error into the standard output.
+        The harness checks the result and will raise an exception if stderr is non-empty.
+
+        Args:
+            container: The specified container or its name.
+            command_prefix: The command prefix to register against.
+            handler: A handler function that simulates the command's execution.
+            result: A simplified form to specify the command's simulated result.
+
+        Example usage::
+
+            # produce no output and return 0 for every command
+            harness.handle_exec('container', [], result=0)
+
+            # simple example that just produces output (exit code 0)
+            harness.handle_exec('webserver', ['ls', '/etc'], result='passwd\nprofile\n')
+
+            # slightly more complex (use stdin)
+            harness.handle_exec(
+                'c1', ['sha1sum'],
+                handler=lambda args: ExecResult(stdout=hashlib.sha1(args.stdin).hexdigest()))
+
+            # more complex example using args.command
+            def docker_handler(args: testing.ExecArgs) -> testing.ExecResult:
+                match args.command:
+                    case ['docker', 'run', image]:
+                        return testing.ExecResult(stdout=f'running {image}')
+                    case ['docker', 'ps']:
+                        return testing.ExecResult(stdout='CONTAINER ID   IMAGE ...')
+                    case _:
+                        return testing.ExecResult(exit_code=1, stderr='unknown command')
+
+            harness.handle_exec('database', ['docker'], handler=docker_handler)
+
+            # handle timeout
+            def handle_timeout(exec_args: testing.ExecArgs) -> int:
+                if exec_args.timeout is not None and exec_args.timeout < 10:
+                    raise TimeoutError
+                return 0
+
+            harness.handle_exec('database', ['foo'], handler=handle_timeout)
+        """
+        if (handler is None and result is None) or (handler is not None and result is not None):
+            raise TypeError("Either handler or result must be provided, but not both.")
+        container_name = container if isinstance(container, str) else container.name
+        if isinstance(result, int):
+            result = ExecResult(exit_code=result)
+        if isinstance(result, (str, bytes)):
+            result = ExecResult(stdout=result)
```

Comment:
How about this, to help people not using static type checking? The "not bool" test is because bools are ints, but we don't want someone confusing things by returning True (1) or False (0).

```suggestion
        if isinstance(result, int) and not isinstance(result, bool):
            result = ExecResult(exit_code=result)
        elif isinstance(result, (str, bytes)):
            result = ExecResult(stdout=result)
        elif not isinstance(result, ExecResult):
            raise TypeError(f"result must be int, str, bytes, or ExecResult, not {result.__class__.__name__}")
```

Reply:
Never know the `bool` is `int`, but I guess it makes sense, updated, thanks!